src/ca.rs

```diff
@@ -166,573 +139,382 @@ tcg-dice-kp-eca = 2.23.133.5.4.100.12
+        // else we'll get back the path to the CSR so that it can be exported
+        // and eventually certified by some external process
+        let (src, dst) = if spec.self_signed {
+            // sleep to let sessions cycle
+            thread::sleep(Duration::from_millis(1500));
+
+            info!("Generating self-signed cert for CA root");
+            let mut cmd = Command::new("openssl");
+            cmd.arg("ca")
+                .arg("-batch")
+                .arg("-selfsign")
+                .arg("-notext")
+                .arg("-config")
+                .arg("openssl.cnf")
+                .arg("-engine")
+                .arg("pkcs11")
+                .arg("-keyform")
+                .arg("engine")
+                .arg("-keyfile")
+                .arg(format!("0:{:04x}", spec.id))
+                .arg("-extensions")
+                .arg(spec.purpose.to_string())
+                .arg("-passin")
+                .arg("env:OKM_HSM_PKCS11_AUTH")
+                .arg("-in")
+                .arg(csr.path())
+                .arg("-out")
+                .arg(CA_CERT)
+                .output()?;
+
+            debug!("executing command: \"{:#?}\"", cmd);
+            let output = match cmd
+                .output()
+                .context("Failed to self sign cert with `openssl ca`")
+            {
+                Ok(o) => o,
+                Err(e) => {
+                    teardown_warn_only(connector, pwd);
+                    return Err(e);
+                }
+            };
+
+            if !output.status.success() {
+                warn!("command failed with status: {}", output.status);
+                warn!(
+                    "stderr: \"{}\"",
+                    String::from_utf8_lossy(&output.stderr)
+                );
+                teardown_warn_only(connector, pwd);
+                return Err(CaError::SelfCertGenFail.into());
+            }
 
-    Ok(())
-}
+            (
+                root.as_ref().to_path_buf().join(CA_CERT),
+                out.join(format!("{}.cert.pem", spec.label)),
+            )
+        } else {
+            // self-signed=false in keyspec indicates that the CA being
+            // initialized is an intermediate: someone else has to certify it
+            // so we copy the CSR to output
+            (
+                csr.path().to_path_buf(),
+                out.join(format!("{}.csr.pem", spec.label)),
+            )
+        };
 
-pub fn sign(
-    spec: &Path,
-    state: &Path,
-    publish: &Path,
-    transport: Transport,
-) -> Result<()> {
-    let spec = fs::canonicalize(spec)?;
-    debug!("canonical spec path: {}", &spec.display());
-
-    let paths = if spec.is_file() {
-        vec![spec.clone()]
-    } else {
-        config::files_with_ext(&spec, CSRSPEC_EXT)?
-            .into_iter()
-            .chain(config::files_with_ext(&spec, DCSRSPEC_EXT)?)
-            .collect::<Vec<PathBuf>>()
-    };
+        fs::copy(&src, &dst).with_context(|| {
```

Comment:
This fails meet the stated goal of getting paths to outputs out of this module. Since this function returns either a CSR or a cert we could use a data bearing enum instead? Something like:

```
// both arms are ASN.1 encoded as DER that is again encoded in a String / PEM
enum CertOrCsr {
    Cert(String),
    Csr(String),
}
```